lib/drivers/medtronic600Driver.js

```diff
@@ -994,6 +999,310 @@ class ReadBasalPatternCommand extends TransmitPacketRequest {
+  }
+
+  processSegment(packetNumber, payload, hidDevice) {
+    // eslint-disable-next-line no-unused-vars
+    return new Promise((resolve, reject) => {
+      debug('*** GOT A MULTIPACKET SEGMENT', packetNumber, ', count:', this.segmentCount);
+
+      // multiByteSegments don't always come back in a consecutive order.
+      this.segments[packetNumber] = payload;
+
+      if (packetNumber === this.lastPacketNumber &&
+        payload.length !== this.lastPacketSize) {
+        reject(new InvalidMessageError('Multipacket Transfer last packet size mismatch'));
+      } else if (packetNumber !== this.lastPacketNumber &&
+        payload.length !== this.packetSize) {
+        reject(new InvalidMessageError('Multipacket Transfer packet size mismatch'));
+      } else if (this.retransmitNeeded()) {
+        debug('*** NEED SOME PACKETS RETRANSMITTED! ***', this.missingSegmentKeys);
+
+        new AckMultipacketCommand(this.pumpSession,
+            AckMultipacketCommand.SEGMENT_COMMAND.REPEAT_SEGMENT_PACKETS)
+          .send(hidDevice)
+          .then(() => resolve());
+      } else if (this.segmentComplete()) {
+        const segmentPayload = this.segmentPayload;
+        if (segmentPayload.length !== this.segmentSize) {
+          reject(new InvalidMessageError('Total segment size mismatch'));
+        }
+
+        // TODO - all of this should go into different classes...
+        // Decompress the message
+        if (segmentPayload.readUInt16BE(0x00) === 0x030E) {
+          const HEADER_SIZE = 12; // TODO should be a static get.
+          // It's an UnmergedHistoryUpdateCompressed response. We need to decompress it
+          const dataType = segmentPayload[0x02]; // Returns a HISTORY_DATA_TYPE
+          const historySizeCompressed = segmentPayload.readUInt32BE(0x03);
+          const historySizeUncompressed = segmentPayload.readUInt32BE(0x07);
+          const historyCompressed = segmentPayload[0x0B];
+
+          if (dataType !== this.historyDataType) {
+            reject(new InvalidMessageError('Unexpected history type in response'));
+          }
+
+          // Check that we have the correct number of bytes in this message
+          if (segmentPayload.length - HEADER_SIZE !== historySizeCompressed) {
+            reject(new InvalidMessageError('Unexpected message size'));
+          }
+
+          let blockPayload = null;
+          if (historyCompressed) {
+            blockPayload = lzo.decompress(segmentPayload.slice(HEADER_SIZE));
+            if (blockPayload.length !== historySizeUncompressed) {
+              reject(new InvalidMessageError('Unexpected uncompressed message size'));
+            }
+          } else {
+            blockPayload = segmentPayload.slice(HEADER_SIZE);
+          }
+
+          if (blockPayload.length % ReadHistoryCommand.BLOCK_SIZE) {
+            reject(new InvalidMessageError('Block payload size is not a multiple of 2048'));
```

Comment:
I definitely still need to add error handling and retries to the download part of the code.